cheroot/connections.py

```diff
@@ -99,76 +99,87 @@ def expire(self):
-            return self._readable_conns.popleft()
-
-        # Will require a select call.
-        try:
-            # The timeout value impacts performance and should be carefully
-            # chosen. Ref:
-            # github.com/cherrypy/cheroot/issues/305#issuecomment-663985165
-            rlist = [
-                key for key, _
-                in self._selector.select(timeout=0.01)
-            ]
-        except OSError:
-            # Mark any connection which no longer appears valid
-            for _, key in self._selector.get_map().items():
-                # If the server socket is invalid, we'll just ignore it and
-                # wait to be shutdown.
-                if key.data == self.server:
-                    continue
+        Connections submitted for processing must be `put()`
+        back if it should be examined again for another request.
 
-                try:
-                    os.fstat(key.fd)
-                except OSError:
-                    # Socket is invalid, close the connection
-                    self._selector.unregister(key.fd)
+        Can be shut down by calling `stop()`.
+        """
+        self._serving = True
+        expiration_interval = 0.5
+        last_expiration_check = time.time()
+
+        while not self._stop_requested:
+            try:
+                rlist = [
+                    key for key, _
+                    in self._selector.select(timeout=expiration_interval)
+                ]
+            except OSError:
+                self._remove_invalid_sockets()
+                continue
+
+            for key in rlist:
+                if key.data is self.server:
+                    # New connection
+                    conn = self._from_server_socket(self.server.socket)
+                    self.server.process_conn(conn)
+                else:
                     conn = key.data
-                    conn.close()
+                    # unregister connection from the selector until the server
+                    # has read from it and returned it via put()
+                    self._selector.unregister(key.fd)
+                    self.server.process_conn(conn)
 
-            # Wait for the next tick to occur.
-            return None
+            now = time.time()
+            if (now - last_expiration_check) > expiration_interval:
+                self._expire()
+                last_expiration_check = now
 
-        for key in rlist:
-            if key.data is self.server:
-                # New connection
-                return self._from_server_socket(self.server.socket)
+        self._serving = False
 
-            conn = key.data
-            # unregister connection from the selector until the server
-            # has read from it and returned it via put()
+    def _remove_invalid_sockets(self):
+        # Mark any connection which no longer appears valid
+        invalid_keys = []
+        for _, key in self._selector.get_map().items():
```

Comment:
```suggestion
        for _, key in self._selector.get_map().copy().items():
```

Reply:
(some articles suggest using `list(dict().items())` but that's wrong because casting to list iterates over a moving target vs `dict().copy().items()` that allows to iterate over some version of dict that is guaranteed to be immutable)